src/vsr/replica.zig

```diff
@@ -3338,6 +3416,19 @@ pub fn Replica(
             return true;
         }
 
+        /// Returns the highest op that can be received, considering the current checkpoint.
+        fn op_ceiling(self: *const Self) u64 {
+            assert(self.op_checkpoint <= self.commit_min);
+            assert(self.op_checkpoint <= self.op);
+            assert((self.op_checkpoint + 1) % config.lsm_batch_multiple == 0);
```

Comment:
Perhaps we could have a `checkpoint_op(op) bool` helper to encapsulate all these?

Reply:
Let's rename to `op_checkpoint_next()` since that's what it is: the next checkpoint op after `op_checkpoint`.

Reply:
Are you proposing that `op_checkpoint_next` replace `op_ceiling`? I think they still need to be separate because when `op_checkpoint=0`, `op_checkpoint_next == journal_slot_count - lsm_batch_multiple`, but `op_ceiling == journal_slot_count - 1` (you're right that `op_ceiling` doesn't need a branch, and has an off-by-one).

Reply:
Nevermind, I think you just meant that the checks should be extracted into `op_checkpoint_next`.

Reply:
Thanks for the explanation, I understand better now.